### ISSUE TITLE
Internal Server Error when opening any tab on cups homepage

### ISSUE THREAD

**Issue by alllexx88** [apple/cups#5188]
Hi,

I'm facing a cups issue that I don't know how to debug. cups web page returns 500 Internal Server Error whenever opening Administration/Classes/Help/Jobs/Printers tabs. I've enabled maximum log verbosity (debug2), but I still see nothing meaningful/important beyond `cupsdSendError code=500, auth_type=0`, which leads me nowhere. Here's the log: https://pastebin.com/nzQrs45L

cups version is `2.2.6`, it's an Optware-ng package on an armv7 router. Optware-ng currently uses patches from Debian (see [here](https://github.com/Optware/Optware-ng/tree/9b483cefff9f9759d3c5041756d0b24af6f04810/sources/cups/debian)) and two local patches: [cupsd-set-default-for-RIPCache-to-auto.patch](https://github.com/Optware/Optware-ng/blob/9b483cefff9f9759d3c5041756d0b24af6f04810/sources/cups/cupsd-set-default-for-RIPCache-to-auto.patch) and [ppdc.patch](https://github.com/Optware/Optware-ng/blob/9b483cefff9f9759d3c5041756d0b24af6f04810/sources/cups/ppdc.patch). I've tried building cups without any patches (except the second local one, needed to cross-compile cups), the result is the same, and, by the way, the error_log from above was generated by an 'almost-clean' cups-2.2.6 version (ppdc.patch only).

The configure options are:
```
--verbose \
--build=x86_64-pc-linux-gnu \
--host=arm-linux \
--target=arm-linux \
--prefix=/opt \
--exec_prefix=/opt \
--with-icondir=/opt/share/icons \
--with-menudir=/opt/share/applications \
--libdir=/opt/lib \
--with-printcap=/opt/etc/printcap \
--with-local-protocols="dnssd" \
--disable-dbus \
--disable-pam \
--with-rundir=/opt/var/run/cups \
--with-logdir=/opt/var/log/cups \
--enable-gnutls \
--without-java \
--without-perl \
--without-php \
--without-python \
--disable-gssapi \
--with-cups-user=nobody \
--with-cups-group=nobody \
--with-system-groups="root sys system" \
--with-lpdconfigfile=xinetd:///opt/etc/xinetd.d/cups-lpd \
--with-smbconfigfile=samba:///opt/etc/samba/smb.conf \
```

Also, the same issue happened with cups `2.2.4`.

Please tell me if there's something else I can do to debug this issue.

Thanks,
Alex

P.S. I forgot to mention another interesting part of the error_log:
```
D [12/Dec/2017:20:37:01 +0200] [Client 2] Waiting for CGI data.
d [12/Dec/2017:20:37:01 +0200] cupsdCheckJobs: 0 active jobs, sleeping=0, ac-power=-1, reload=0, curtime=1513103821
d [12/Dec/2017:20:37:01 +0200] process_children()
d [12/Dec/2017:20:37:01 +0200] cupsdFinishProcess(pid=24546, name=0xbe874398, namelen=1024, job_id=0xbe87436c(0)) = "/opt/lib/cups/cgi-bin/jobs.cgi"
D [12/Dec/2017:20:37:01 +0200] PID 24546 (/opt/lib/cups/cgi-bin/jobs.cgi) stopped with status 127 (File too large)
```

This at least gives the cgi crash error code. However, I don't know where to dig further...

**Comment by alllexx88** [comment]
I got this issue fixed. It's the same issue with uClibc-ng as #4496 with OpenBSD. Here's the patch that fixes this issue for me:

```
--- a/scheduler/process.c
+++ b/scheduler/process.c
@@ -25,7 +25,7 @@
 extern char **environ;
 #endif /* HAVE_POSIX_SPAWN */
 #ifdef HAVE_POSIX_SPAWN
-#  if !defined(__OpenBSD__) || OpenBSD >= 201505
+#  if (!defined(__OpenBSD__) || OpenBSD >= 201505) && !defined(__UCLIBC__)
 #    define USE_POSIX_SPAWN 1
 #  else
 #    define USE_POSIX_SPAWN 0
```

**Comment by michaelrsweet** [comment]
@alllexx88 Have you reported this bug in posix_spawn to the uClibc folks?

I hesitate to unilaterally disable posix_spawn with uClibc - seems like we either want a similar version check or add a configure option to disable posix_spawn.

**Comment by alllexx88** [comment]
@michaelrsweet I'll do this soon, now that I've narrowed it down to a standalone working example that works with GNU lib C, but fails with uClibc-mg 1.0.26:

```
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <unistd.h>
#include <spawn.h>
#include <sys/wait.h>

extern char **environ;

void run_cmd(char *cmd)
{
    pid_t pid;
    posix_spawnattr_t attrs;
    posix_spawn_file_actions_t actions;
    sigset_t defsignals;
    char *argv[] = {"sh", "-c", cmd, NULL};
    int status;

    sigemptyset(&defsignals);
    sigaddset(&defsignals, SIGTERM);
    sigaddset(&defsignals, SIGCHLD);
    sigaddset(&defsignals, SIGPIPE);

    posix_spawnattr_init(&attrs);
    posix_spawnattr_setflags(&attrs, POSIX_SPAWN_SETPGROUP | POSIX_SPAWN_SETSIGDEF);
    posix_spawnattr_setpgroup(&attrs, 0);
    posix_spawnattr_setsigdefault(&attrs, &defsignals);

    printf("Run command: %s\n", cmd);
    status = posix_spawn(&pid, "/bin/sh", &actions, &attrs, argv, environ);
    if (status == 0) {
        printf("Child pid: %i\n", pid);
        if (waitpid(pid, &status, 0) != -1) {
            printf("Child exited with status %i\n", status);
        } else {
            perror("waitpid");
        }
    } else {
        printf("posix_spawn: %s\n", strerror(status));
    }
}

int main(int argc, char* argv[])
{
    run_cmd(argv[1]);
    return 0;
}
```

Just like it was with OpenBSD, the problems seems to be with `POSIX_SPAWN_SETSIGDEF` usage. If I remove this flag, `posix_spawn` works fine.

**Comment by alllexx88** [comment]
@michaelrsweet Looks like posix_spawn works fine starting from uClibc-ng version 1.0.27 (Optware-ng used 1.0.26, now upgraded to 1.0.27). We could test for uClibc version like this:

```
--- a/scheduler/process.c
+++ b/scheduler/process.c
@@ -25,11 +25,15 @@
 extern char **environ;
 #endif /* HAVE_POSIX_SPAWN */
 #ifdef HAVE_POSIX_SPAWN
-#  if !defined(__OpenBSD__) || OpenBSD >= 201505
+#  if (!defined(__OpenBSD__) || OpenBSD >= 201505) && \
+      (!defined(__UCLIBC__) || (__UCLIBC_MAJOR__ > 1) || \
+       ((__UCLIBC_MAJOR__ == 1) && (__UCLIBC_MINOR__ > 0)) || \
+       ((__UCLIBC_MAJOR__ == 1) && (__UCLIBC_MINOR__ == 0) && \
+                                   (__UCLIBC_SUBLEVEL__ >= 27))
 #    define USE_POSIX_SPAWN 1
 #  else
 #    define USE_POSIX_SPAWN 0
-#  endif /* !__OpenBSD__ || */
+#  endif /* (!defined(__OpenBSD__) || */
 #else
 #  define USE_POSIX_SPAWN 0
 #endif /* HAVE_POSIX_SPAWN */
```

Looks a bit cluttered to me, but it gets the job done. What do you think about it?

**Comment by michaelrsweet** [comment]
[master 9e44dfbe6] Add conditional code for posix_spawn when using uClibc (Issue #5188)

**Comment by mschubert** [comment]
If anyone finds this bug thinking it is not fixed in `2.2.7`, make sure the */etc/cups* is owned by `cups` and *cups-files.conf* has values `User` and `Group` set to `209`.
